File: go/muse/data_structures/singly_linked_list/singly_linked_list.go
package singly_linked_list

import "errors"

type Node[T any] struct {
	data T
	next *Node[T]
}

type SinglyLinkedList[T any] struct {
	head   *Node[T]
	length int
}

func New[T any]() *SinglyLinkedList[T] {
	return &SinglyLinkedList[T]{head: nil, length: 0}
}

func (l *SinglyLinkedList[T]) Size() int {
	return l.length
}

func (l *SinglyLinkedList[T]) IsEmpty() bool {
	return l.length == 0
}

func (l *SinglyLinkedList[T]) Get(index int) T {
	if index < 0 || index >= l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	cursor := l.head

	for range index {
		cursor = cursor.next
	}

	return cursor.data
}

func (l *SinglyLinkedList[T]) Set(index int, element T) {
	if index < 0 || index >= l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	cursor := l.head

	for range index {
		cursor = cursor.next
	}

	cursor.data = element
}

func (l *SinglyLinkedList[T]) Insert(index int, element T) {
	if index < 0 || index > l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	node := &Node[T]{data: element, next: nil}

	if index == 0 {
		if l.length != 0 {
			node.next = l.head
		}
		l.head = node
	} else {
		cursor := l.head
		for range index - 1 {
			cursor = cursor.next
		}
		node.next = cursor.next
		cursor.next = node
	}

	l.length++
}

func (l *SinglyLinkedList[T]) Remove(index int) {
	if index < 0 || index >= l.length {
		panic(errors.New("[PANIC - IndexOutOfBounds]"))
	}

	var target *Node[T]

	if index == 0 {
		target = l.head
		if l.length == 1 {
			l.head = nil
		} else {
			l.head = target.next
		}
	} else {
		cursor := l.head
		for range index - 1 {
			cursor = cursor.next
		}
		target = cursor.next
		cursor.next = target.next
	}

	target.data = *new(T)

	l.length--
}

func (l *SinglyLinkedList[T]) Front() T {
	return l.Get(0)
}

func (l *SinglyLinkedList[T]) Back() T {
	return l.Get(l.length - 1)
}

func (l *SinglyLinkedList[T]) Prepend(element T) {
	l.Insert(0, element)
}

func (l *SinglyLinkedList[T]) Append(element T) {
	l.Insert(l.length, element)
}

func (l *SinglyLinkedList[T]) Poll() {
	l.Remove(0)
}

func (l *SinglyLinkedList[T]) Eject() {
	l.Remove(l.length - 1)
}

File: rust/muse/src/data_structures/singly_linked_list.rs
use super::List;

#[derive(Debug)]
pub struct Node<T> {
    data: Option<T>,
    next: Option<Box<Node<T>>>,
}

#[derive(Debug)]
pub struct SinglyLinkedList<T> {
    head: Option<Box<Node<T>>>,
    length: usize,
}

impl<T> SinglyLinkedList<T> {
    pub fn new() -> Self {
        Self {
            head: None,
            length: 0,
        }
    }
}

impl<T> Default for SinglyLinkedList<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T> List<T> for SinglyLinkedList<T> {
    fn size(&self) -> usize {
        self.length
    }

    fn is_empty(&self) -> bool {
        self.length == 0
    }

    fn get(&self, index: usize) -> &T {
        if index >= self.length {
            panic!("[PANIC - IndexOutOfBounds]");
        }

        let mut cursor = &self.head;

        (0..index).for_each(|_| {
            cursor = &cursor.as_ref().expect("element should exist").next;
        });

        cursor
            .as_ref()
            .expect("element should exist")
            .data
            .as_ref()
            .expect("element should exist")
    }

    fn set(&mut self, index: usize, element: T) {
        if index >= self.length {
            panic!("[PANIC - IndexOutOfBounds]");
        }

        let mut cursor = &mut self.head;

        for _ in 0..index {
            cursor = &mut cursor.as_mut().expect("element should exist").next;
        }

        cursor.as_mut().expect("element should exist").data = Some(element);
    }

    fn insert(&mut self, index: usize, element: T) {
        if index > self.length {
            panic!("[PANIC - IndexOutOfBounds]");
        }

        let mut node = Box::new(Node {
            data: Some(element),
            next: None,
        });

        if index == 0 {
            if self.length != 0 {
                node.next = self.head.take();
            }
            self.head = Some(node);
        } else {
            let mut cursor = &mut self.head;
            for _ in 0..index - 1 {
                cursor = &mut cursor.as_mut().expect("element should exist").next;
            }
            node.next = cursor.as_mut().expect("element should exist").next.take();
            cursor.as_mut().expect("element should exist").next = Some(node);
        }

        self.length += 1;
    }

    fn remove(&mut self, index: usize) {
        if index >= self.length {
            panic!("[PANIC - IndexOutOfBounds]");
        }

        if index == 0 {
            self.head
                .as_mut()
                .expect("element should exist")
                .data
                .take();
            if self.length == 1 {
                self.head.take();
            } else {
                self.head = self
                    .head
                    .as_mut()
                    .expect("element should exist")
                    .next
                    .take();
            }
        } else {
            let mut cursor = &mut self.head;
            for _ in 0..index - 1 {
                cursor = &mut cursor.as_mut().expect("element should exist").next;
            }
            let target = &mut cursor.as_mut().expect("element should exist").next;
            target.as_mut().expect("element should exist").data.take();
            cursor.as_mut().expect("element should exist").next =
                target.as_mut().expect("element should exist").next.take();
        }

        self.length -= 1;
    }

    fn front(&self) -> &T {
        self.get(0)
    }

    fn back(&self) -> &T {
        self.get(self.length - 1)
    }

    fn prepend(&mut self, element: T) {
        self.insert(0, element)
    }

    fn append(&mut self, element: T) {
        self.insert(self.length, element)
    }

    fn poll(&mut self) {
        self.remove(0)
    }

    fn eject(&mut self) {
        self.remove(self.length - 1)
    }
}

#[cfg(test)]
pub(crate) mod tests;
